// @nhost/nhost-js/dist/nhost-js/fetch.js
import { F, c } from "../fetch-2PHmQBIJ.js";
import { a, s, u, w } from "../middlewareWithAdminSession-BeIk-9HO.js";
const withHeadersMiddleware = (defaultHeaders) => (next) => async (url, options = {}) => {
  const headers = new Headers(options.headers || {});
  const defaults = new Headers(defaultHeaders);
  defaults.forEach((value, key) => {
    if (!headers.has(key)) {
      headers.set(key, value);
    }
  });
  return next(url, { ...options, headers });
};
const withRoleMiddleware = (role) => (next) => async (url, requestOptions = {}) => {
  const headers = new Headers(requestOptions.headers || {});
  if (!headers.has("x-hasura-role")) {
    headers.set("x-hasura-role", role);
  }
  return next(url, { ...requestOptions, headers });
};
export {
  F as FetchError,
  a as attachAccessTokenMiddleware,
  c as createEnhancedFetch,
  s as sessionRefreshMiddleware,
  u as updateSessionFromResponseMiddleware,
  w as withAdminSessionMiddleware,
  withHeadersMiddleware,
  withRoleMiddleware
};
