// @nhost/nhost-js/dist/nhost-js/functions.js
import { c as createEnhancedFetch, F as FetchError } from "../fetch-2PHmQBIJ.js";
const createAPIClient = (baseURL, chainFunctions = []) => {
  let enhancedFetch = createEnhancedFetch(chainFunctions);
  const pushChainFunction = (chainFunction) => {
    chainFunctions.push(chainFunction);
    enhancedFetch = createEnhancedFetch(chainFunctions);
  };
  const fetch = async (path, options) => {
    const resp = await enhancedFetch(`${baseURL}${path}`, options);
    let body;
    if (resp.headers.get("content-type")?.includes("application/json")) {
      body = await resp.json();
    } else if (resp.headers.get("content-type")?.startsWith("text/")) {
      body = await resp.text();
    } else {
      body = await resp.blob();
    }
    if (!resp.ok) {
      throw new FetchError(body, resp.status, resp.headers);
    }
    return {
      status: resp.status,
      body,
      headers: resp.headers
    };
  };
  const post = async (path, body, options = {}) => {
    const requestOptions = {
      ...options,
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        ...options.headers
      },
      body: body ? JSON.stringify(body) : void 0
    };
    return fetch(path, requestOptions);
  };
  return {
    baseURL,
    fetch,
    post,
    pushChainFunction
  };
};
export {
  createAPIClient
};
