// @nhost/nhost-js/dist/nhost-js/graphql.js
import { c as createEnhancedFetch, F as FetchError } from "../fetch-2PHmQBIJ.js";
const createAPIClient = (url, chainFunctions = []) => {
  let enhancedFetch = createEnhancedFetch(chainFunctions);
  const pushChainFunction = (chainFunction) => {
    chainFunctions.push(chainFunction);
    enhancedFetch = createEnhancedFetch(chainFunctions);
  };
  const executeOperation = async (request2, options) => {
    const response = await enhancedFetch(`${url}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json"
      },
      body: JSON.stringify(request2),
      ...options
    });
    const body = await response.text();
    const data = body ? JSON.parse(body) : {};
    const resp = {
      body: data,
      status: response.status,
      headers: response.headers
    };
    if (data.errors) {
      throw new FetchError(data, response.status, response.headers);
    }
    return resp;
  };
  function request(requestOrDocument, variablesOrOptions, options) {
    if (typeof requestOrDocument === "object" && "kind" in requestOrDocument) {
      const definition = requestOrDocument.definitions[0];
      const request2 = {
        query: requestOrDocument.loc?.source.body || "",
        variables: variablesOrOptions,
        operationName: definition && "name" in definition ? definition.name?.value : void 0
      };
      return executeOperation(request2, options);
    } else {
      const request2 = requestOrDocument;
      const requestOptions = variablesOrOptions;
      return executeOperation(request2, requestOptions);
    }
  }
  return {
    request,
    url,
    pushChainFunction
  };
};
export {
  createAPIClient
};
